### ISSUE TITLE
make_self with private constructor

### ISSUE THREAD

**Issue by benstevens48** [MicrosoftDocs/winrt-related#79]
Is it possible to create a class with a private constructor (designed to be created via a static method) and call make_self to create a WinRT component instance of the class? I can't get that to work. If this is not possible, is there another way to do it?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 35118798-8f23-27b0-8d63-642dba40b0a0
* Version Independent ID: e56d787a-1de8-422c-fed4-c0e257611b77
* Content: [winrt::make_self function template (C++/WinRT) - Windows UWP applications](https://docs.microsoft.com/en-us/uwp/cpp-ref-for-winrt/make-self)
* Content Source: [winrt-related-src/cpp-ref-for-winrt/make-self.md](https://github.com/MicrosoftDocs/winrt-related/blob/docs/winrt-related-src/cpp-ref-for-winrt/make-self.md)
* Product: **uwp**
* GitHub Login: @stevewhims
* Microsoft Alias: **stwhi**

**Comment by stevewhims** [comment]
Hi, @benstevens48, yes if you omit a default constructor from your IDL, then the default constructor will be deleted in the generated C++ class (FWIW, you're then free to add a default constructor, or any other kind of constructor, in C++ as part of your implementation). Then, in your static factory method, you can use either make or make_self. They both call a constructor on your implementation type, forwarding any arguments. Your factory method will return the projected type (remember that MyProject::MyRuntimeClass is the projected type, and MyProject::implementation::MyRuntimeClass is the implementation type). Use make if you can, and make_self if you need to. For more info on when you might need to, see https://docs.microsoft.com/windows/uwp/cpp-and-winrt-apis/author-apis#instantiating-and-returning-implementation-types-and-interfaces and search for the string "If you need to access all of the implementation's members, and then later return an interface to a caller". I would recommend studying that topic closely, because there are a lot of details and useful knowledge in there. ~Steve

**Comment by benstevens48** [comment]
Hi @stevewhims, thanks for the reply. I was actually trying to make the constructor private, in addition to omitting a default constructor from my IDL. The problem is I then get an error saying make_self cannot access it. Is there any way around this, or is this not a recommended design pattern (by the way, my C++ knowledge is a bit patchy)?

**Comment by stevewhims** [comment]
I see. I'm thinking, then, that there are two notions of "private" in the C++/WinRT world (or, more generally, in the COM world using C++). There's the "do you want your COM type to be directly creatable publicly?" and "do you want your implementation type to be directly creatable publicly (via a constructor)?". If the answer to the first question is "no", then we know already about that (don't put a constructor in your IDL, but instead have the COM type creatable indirectly via some factory method or via some other type). That first question is a pure COM/WinRT question, and not related to the programming language you use to implement the COM type. However, the second question is a pure C++ one. The reason you might be inclined to answer "no" to that (and hence make your constructor private) is to prevent yourself from accidentally calling that constructor *in the project that implements the COM type* (naturally, no such accident can occur from outside that project). And instead, force yourself to instantiate indirectly via a public factory method and a private constructor. However, a private constructor (and winrt::make and make_self) is not possible with the C++ classes that the C++/WinRT tooling generates for you by default. You could jettison those default classes and write your own (they're only there if you wish to use them) but I'm not sure I see a huge advantage to the private constructor in this case (since, again, the primary consumer of the type is external to the implementing project, and you already have the safety there of not having a constructor in the IDL). You can still preserve invariants, take arguments, and ensure proper initialization in a non-private constructor, and then call that from your factory method while also knowing that it's not the end of the world if some other part of your code bypasses that factory and returns the projected interface via some other means.

In summary, I would (for expediency and convenience) stick with the generated C++ class even though it means my constructors (default and otherwise) can't be private. And I'd implement a factory method that returns the projected type, and I'd call winrt::make from that (which will in turn call my constructor). If I've misinterpreted your reason for wanting your constructor to be private, then I'd be interested to know more about that.

Thanks! :)
~Steve

**Comment by benstevens48** [comment]
Hi Steve, 

Thanks for the reply - it's good to have that clarified. Yes, the reason for making the constructor private was basically to remind myself that the intended way to construct a class instance is through the static factory function, and because I felt like this might be best practice. But I agree that the primary consumer is external and it's not a big problem for the constructor to be public within the implementation, so I will follow your suggestion.

In general I'm finding C++/WinRT a nice way of bridging COM and a UWP C# app, the main downsides being the lack of examples on the web (hence this question in the documentation since it was vaguely related), and the difficulty of finding the relevant error message if you do something that causes the C++/WinRT tooling not to be able to generate all its output properly, since it causes lots of seemingly unrelated errors!

**Comment by stevewhims** [comment]
Sounds good! I'll go ahead and close this GitHub issue. Thanks!
